linera-rpc/src/simple/transport.rs

```diff
@@ -186,52 +192,111 @@ impl ConnectionPool for UdpConnectionPool {
     }
 }
 
-// Server implementation for UDP.
-impl TransportProtocol {
-    async fn run_udp_server<S>(address: impl ToSocketAddrs, state: S) -> Result<(), std::io::Error>
-    where
-        S: MessageHandler + Send + 'static,
-    {
-        let socket = UdpSocket::bind(address).await?;
-        let (udp_sink, mut udp_stream) = UdpFramed::new(socket, Codec).split();
-        let udp_sink = Arc::new(Mutex::new(udp_sink));
-        // Track the latest tasks for a given peer. This is used to return answers in the
-        // same order as the queries.
-        let mut previous_tasks = HashMap::new();
-
-        while let Some(value) = udp_stream.next().await {
-            let (message, peer) = match value {
-                Ok(value) => value,
-                Err(codec::Error::Io(io_error)) => return Err(io_error),
-                Err(other_error) => {
-                    warn!("Received an invalid message: {other_error}");
-                    continue;
+/// Server implementation for UDP.
+pub struct UdpServer<State> {
+    handler: State,
+    udp_sink: SharedUdpSink,
+    udp_stream: SplitStream<UdpFramed<Codec>>,
+    active_handlers: HashMap<SocketAddr, JoinHandle<()>>,
+}
+
+/// Type alias for the outgoing endpoint of UDP messages.
+type SharedUdpSink = Arc<Mutex<SplitSink<UdpFramed<Codec>, (RpcMessage, SocketAddr)>>>;
+
+impl<State> UdpServer<State>
+where
+    State: MessageHandler + Send + 'static,
+{
+    /// Runs the UDP server implementation.
+    pub async fn run(
+        address: impl ToSocketAddrs,
+        state: State,
+        shutdown_signal: CancellationToken,
+    ) -> Result<(), std::io::Error> {
+        let mut server = Self::bind(address, state).await?;
+
+        loop {
+            tokio::select! { biased;
```

Comment:
What is this `biased; ` ? Is there a Rust way to write this?

Reply:
It's to ensure that the `select` arms are polled in order, instead of pseudo-randomly. This is to ensure that it stops as soon as the cancellation request is received. I don't think there's another way to do this with Tokio's `select!`, because this is the [recommended way](https://docs.rs/tokio/latest/tokio/macro.select.html#fairness).